### PR TITLE
Fix: enable payer to create a second stream for the same payee

### LINKED ISSUE #81: Problem: can't create a second stream for the same payee after terminating
```
if (
            streamInstances[_payee].createdAt > 0 &&
            streamInstances[_payee].terminatedAt == 0
        ) revert OpenStreamExists();
        if (
            streamInstances[_payee].createdAt > 0 &&
            streamInstances[_payee].terminatedAt != 0 &&
            streamInstances[_payee].createdAt + streamInstances[_payee].cliffPeriod < streamInstances[_payee].terminatedAt &&
            streamInstances[_payee].terminatedAt + streamInstances[_payee].terminationPeriod > block.timestamp
        ) revert StreamIsTerminating();
```
When creating a second stream for the sampe payee, it already has createdAt which is greater than zero. So it will be reverted.

### PR DESCRIPTION
**Pull request by accompany1205**
resolves #81 
The payer can create a second stream for the same payee if the payee already claimed everything in the first stream.
For this, current time is behind of termination period + terminated time.

### CODE REVIEW

**Comment by accompany1205** [comment]
@markholdex 
The first stream should be terminated and fully claimed. If no accumulation, possible

**Comment by markholdex** [comment]
Got it. Will review now

**Comment by accompany1205** [comment]
@markholdex I am testing the contracts on testnet. Please review it.

File: contracts/StreamManager.sol [hunk 1]
```diff
@@ -238,6 +208,9 @@ contract StreamManager is IStreamManager, ReentrancyGuard {
         uint256 terminatedAt = block.timestamp;
         if (!isPayee[_payee]) revert NotPayee();
         if (streamInstances[_payee].terminatedAt != 0) revert AlreadyTerminatedOrTerminating();
+        /// Terminte in cliff period
```
Comment (georgeciubotaru) [inline contracts/StreamManager.sol:211]: typo here

Comment (georgeciubotaru) [reply contracts/StreamManager.sol:211]: Better here to get the storage value and modify it

`OpenStream storage streamInstance = streamInstances[_payee];`

**Comment by accompany1205** [comment]
@markholdex I updated

File: contracts/StreamManager.sol [hunk 2]
```diff
@@ -196,25 +168,23 @@ contract StreamManager is IStreamManager, ReentrancyGuard {
...
         uint256 claimedAt = block.timestamp;
         uint256 terminatedAt = streamInstances[msg.sender].terminatedAt;
-        uint256 lastClaimedAt = streamInstances[msg.sender].lastClaimedAt;
         address token = streamInstances[msg.sender].token;
         uint256 terminationPeriod = streamInstances[msg.sender].terminationPeriod;
         bool isTerminated = streamInstances[msg.sender].isTerminated;
         uint256 claimableAmount;
 
-        if (!isTerminated || isTerminated && claimedAt <= terminatedAt + terminationPeriod) {
+        if (!isTerminated || isTerminated && claimedAt < terminatedAt + terminationPeriod) {
```
Comment (georgeciubotaru) [inline contracts/StreamManager.sol:182]: combining || and && should be done like that a || (b && c) of (a || b) && c

**Review by georgeciubotaru — CHANGES_REQUESTED** [review]
see my change.

Also, don't forget if you add new features or remove them you need to add/remove tests accordingly

File: contracts/StreamManager.sol (hunk 2 above)
Comment (georgeciubotaru) [reply contracts/StreamManager.sol:182]: https://github.com/holdex/payment-streaming/pull/88#discussion_r1269623174

File: contracts/StreamManager.sol (hunk 1 above)
Comment (georgeciubotaru) [reply contracts/StreamManager.sol]: yes

Comment (accompany1205) [reply contracts/StreamManager.sol]: Fixed

File: contracts/StreamManager.sol (hunk 2 above)
Comment (accompany1205) [reply contracts/StreamManager.sol]: Yeah

File: contracts/StreamManager.sol [hunk 3]
```diff
@@ -201,25 +173,24 @@ contract StreamManager is IStreamManager, ReentrancyGuard {
...
         nonReentrant
         external
     {
         uint256 claimedAt = block.timestamp;
-        uint256 terminatedAt = streamInstances[msg.sender].terminatedAt;
-        uint256 lastClaimedAt = streamInstances[msg.sender].lastClaimedAt;
-        address token = streamInstances[msg.sender].token;
-        uint256 terminationPeriod = streamInstances[msg.sender].terminationPeriod;
-        bool isTerminated = streamInstances[msg.sender].isTerminated;
+        OpenStream storage streamInstance = streamInstances[msg.sender];
```
Comment (georgeciubotaru) [inline contracts/StreamManager.sol:181]: @accompany1205 use that in all functions

File: test/StreamManager.test.js
Comment (georgeciubotaru) [inline test/StreamManager.test.js:1]: @accompany1205 Don't forget to add test cases for the new lines added

File: contracts/StreamManager.sol (hunk 3 above)
Comment (accompany1205) [reply contracts/StreamManager.sol:181]: Fixed

File: test/StreamManager.test.js
Comment (accompany1205) [reply test/StreamManager.test.js:1]: All done for now. nothing to add. I checked it again.
I rather removed redundant test cases after fixing the contract.

File: contracts/StreamManager.sol [hunk 4]
```diff
@@ -241,8 +212,12 @@ contract StreamManager is IStreamManager, ReentrancyGuard {
...
         uint256 terminatedAt = block.timestamp;
+        OpenStream storage streamInstance = streamInstances[_payee];
         if (!isPayee[_payee]) revert NotPayee();
-        if (streamInstances[_payee].terminatedAt != 0) revert AlreadyTerminatedOrTerminating();
+        if (streamInstance.terminatedAt != 0) revert AlreadyTerminatedOrTerminating();
+        /// Terminate in cliff period
+        if (streamInstance.createdAt + streamInstance.cliffPeriod >= block.timestamp)
+            isPayee[_payee] = false;
         streamInstances[_payee].isTerminated = true;
         streamInstances[_payee].terminatedAt = terminatedAt;
```
Comment (georgeciubotaru) [inline contracts/StreamManager.sol:222]: @accompany1205 you already have the reference of the instance, just update it accordingly

Comment (accompany1205) [reply contracts/StreamManager.sol]: Fixed

**Review by georgeciubotaru — DISMISSED** [review]

**Comment by markholdex** [comment]
@accompany1205 please fix conflicts and let's merge.

**Review by markholdex — APPROVED** [review]

**Pull request merged**